third_party/move/move-compiler-v2/src/pipeline/flush_writes_processor.rs

```diff
@@ -271,12 +314,123 @@ impl FlushWritesProcessor {
                 .get(&prev_use_at_usage_instr)
                 .map_or(false, |defs| {
                     defs.iter().any(|defs_of_prev_use| {
-                        defs_of_prev_use > &def && defs_of_prev_use.offset < *use_offset
+                        defs_of_prev_use > def && defs_of_prev_use.offset < *use_offset
                     })
                 })
         })
     }
 
+    /// Check for various conditions where between a `def` and its `use_`, there are other
+    /// definitions that could cause `def` to be flushed before its `use_`.
+    fn has_flush_causing_defs_in_between(
+        def: &DefOrUsePoint,
+        use_: &DefOrUsePoint,
+        code: &[Bytecode],
+        use_def_links: &UseDefLinks,
+        flush_writes: &BTreeMap<CodeOffset, BTreeSet<TempIndex>>,
+        assign_optimize: bool,
+    ) -> bool {
+        if !assign_optimize {
+            return false;
+        }
+        // For each definition in between `def` and `use_`, is there at least one that is:
```

Comment:
This is still not clear, even after re-reading the comment at top of file.  It seems to have assumptions about the state of the world (like, that code is always processed bottom-up) that should be documented.

Suppose I have:
```
    def x   // "def"
    def y
    ...
    StLoc y  // flush y (StLoc)
    use x   // "use_"
    MoveLoc y
    use y
```
It looks like "def could be flushed before its use_, so we should instead flush it right away."  But there is no need to flush def here, as y is already being flushed, just not "right away".

Assuming that "def y" is processed before "def x", then we would already have a "flush y" right after its def so condition 1 below would apply.

BTW, the "used in the same offset as use_" seems potentially ambiguous/underdetermined:  is there a difference between `use(x, y)` and `use(y, x)` ?

Reply:
> that code is always processed bottom-up that should be documented.

Done.

> BTW, the "used in the same offset as use_" seems potentially ambiguous/underdetermined: is there a difference between use(x, y) and use(y, x) ?

Yes, but we already check in `has_intervening_def` the order within the same offset.

> It looks like "def could be flushed before its use_, so we should instead flush it right away." But there is no need to flush def here, as y is already being flushed, just not "right away".

Yes, there is no need to flush "def", because y is consumed before/flushed before "use_". There is no `StLoc` yet, so the actual alternative is for your code is one of (a) y is marked to be flushed right away, or (b) y is consumed from the stack, due to a call that consumes it, or (c) y ends up being flushed later but before "use_", but we cannot detect it (in which case, we could have a wasted flush).